nxp: Add SAI driver with the DAI interface

LaurentiuM1234 opened the pull request:
For now, this PR shall remain a draft until all NXP HAL PRs are ready.

An example DTS node can be found below (i.MX93, SOF usage):

```
sai3: dai@42660000 {
    compatible = "nxp,dai-sai";
    reg = <0x42660000 DT_SIZE_K(64)>;
    mclk-is-output;
    clocks = <&ccm IMX_CCM_SAI3_CLK 0x0 0x0>;
    clock-names = "mclk1";
    rx-fifo-watermark = <96>;
    tx-fifo-watermark = <1>;
    fifo-depth = <96>;
    interrupt-parent = <&gic>;
    interrupts = <GIC_SPI 171 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
 };
```

The additional configurations required to get the driver working are (i.MX93):

```
CONFIG_DAI=y
CONFIG_SAI_HAS_MCLK_CONFIG_OPTION=y
CONFIG_SAI_HAS_XCR4_CHMOD=y
```

Tagging @dbaluta and @iuliana-prodan for initial comments.

LaurentiuM1234 committed:
drivers: dai: Add driver for NXP's SAI

This commit introduces a new DAI driver used for NXP'S SAI IP.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
clock_control: imx_ccm: Add support for i.MX93's SAI clocks

This commit introduces support for querying
i.MX93's SAI clocks.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
include: dai.h: Add definitions for parsing dai_config's format field

This commit introduces some macros and enums which can
be used to parse struct dai_config's format field. This is
required by the SAI driver since it uses dai_config's format
field to select the protocol, clock configuration and clock
inversion. This is added to the dai.h header to avoid having
to define these macros/enums in each of the DAI drivers.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
manifest: Bump up hal_nxp revision

Bump up hal_nxp revision to contain the latest SAI-related
changes from NXP HAL side.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

danieldegrasse reviewed (commented):
A very broad comment- based on the `dai_protocol` enum, it sort of looks like these formats are just the same as those we support in I2S today: https://docs.zephyrproject.org/latest/hardware/peripherals/audio/i2s.html. Given that, is there any reason this DAI driver could not be written to use the I2S API itself, instead of the SAI peripheral? It seems like the DAI API is intended as a superset of the I2S API, but maybe I'm missing some advantage here.

LaurentiuM1234 commented:
> A very broad comment- based on the `dai_protocol` enum, it sort of looks like these formats are just the same as those we support in I2S today: https://docs.zephyrproject.org/latest/hardware/peripherals/audio/i2s.html. Given that, is there any reason this DAI driver could not be written to use the I2S API itself, instead of the SAI peripheral? It seems like the DAI API is intended as a superset of the I2S API, but maybe I'm missing some advantage here.

Can't use I2S interface as SOF currently uses the DAI interface

dbaluta commented:
> A very broad comment- based on the `dai_protocol` enum, it sort of looks like these formats are just the same as those we support in I2S today: https://docs.zephyrproject.org/latest/hardware/peripherals/audio/i2s.html. Given that, is there any reason this DAI driver could not be written to use the I2S API itself, instead of the SAI peripheral? It seems like the DAI API is intended as a superset of the I2S API, but maybe I'm missing some advantage here.

@danieldegrasse There was a discussion in TSC where Intel guys introduced new subsystem called `DAI`' because already existing `I2S` subsystem didn't fit SOF usecase.

@lgirdwood  can offer more details.

kv2019i commented:
For some background, we have multiple types of audio interfaces behind the DAI interface (I2S, Soundwire and HDA audio) and code in SOF side is using Zephyr DAI interface to have common code to talk to all types of interfaces. This PR allows the NXP SAI Zephyr (DAI) driver to be used in SOF code. If we'd use I2S interface here, SOF application code would require a target/hw based separate code path to talk to I2S driver interface in case of this NXP target, and this is what we wanted to avoid. We have multiple vendors and multiple hw using the DAI interface and we hope to get these all to Zephyr (code already exists, just in SOF), not just for Intel hardware, so I'm glad to see this PR.

danieldegrasse commented:
> If we'd use I2S interface here, SOF application code would require a target/hw based separate code path to talk to I2S driver interface in case of this NXP target, and this is what we wanted to avoid

Absolutely- I don't think we should change SOF application code. What I'm wondering is if it would be possible to create a DAI driver (say `dai_i2s.c`) that implements the DAI interface by calling the I2S APIs, essentially as a wrapper layer. This way there is no change needed to the SOF application layer, and we reuse the existing code in I2S drivers.

dbaluta commented:
> Absolutely- I don't think we should change SOF application code. What I'm wondering is if it would be possible to create a DAI driver (say `dai_i2s.c`) that implements the DAI interface by calling the I2S APIs, essentially as a wrapper layer. This way there is no change needed to the SOF application layer, and we reuse the existing code in I2S drivers.

Although an interesting idea I think that would be an overkill. Some times there will be functions that could be imported from i2s but some cannot. So, I think we complicate things unnecessary. Maybe after we get this in and have it in a stable form we could reduce code duplication by looking at this.

danieldegrasse commented:
> Although an interesting idea I think that would be an overkill. Sometimes there will be functions that could be imported from i2s but sometimes not. So, I think we complicate things unnecessary. Maybe after we get this in and have it in a stable form we could reduce code duplication by looking at this.

Understood- you're the expert here, this was just a passing thought :) If it doesn't make sense than no worries

LaurentiuM1234 commented on drivers/dai/CMakeLists.txt:
```diff
@@ -4,3 +4,5 @@ add_subdirectory_ifdef(CONFIG_DAI_INTEL_SSP intel/ssp)
 add_subdirectory_ifdef(CONFIG_DAI_INTEL_DMIC intel/dmic)
 add_subdirectory_ifdef(CONFIG_DAI_INTEL_HDA intel/hda)
+
+add_subdirectory(nxp)
```
Can't conditionally add `nxp/` as that would mean introducing a new configuration (e.g: `CONFIG_DAI_NXP`) which would be pointless (or, worse, conditionally adding `nxp/` and inside `nxp/`, conditionally compiling the sources). The difference between us and Intel here is the fact that our `nxp/` hierarchy will directly contain the sources for our DAIs.

LaurentiuM1234 replied:
We could go for this, sure, but before that I'd like to understand what benefits that would bring. My knowledge of the CMAKE build system is very modest at best so I'm not sure why we'd choose one approach over the other. Also, when more NXP DAI drivers are introduced won't that require more `add_subdirectory_ifdef()` statements (or a huge one at best)?

iuliana-prodan replied:
Now, the `sai.c` is added, by default, for compilation - of course, if CONFIG_DAI is selected, based on [this](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/CMakeLists.txt#L31).

Using `add_subdirectory_ifdef(CONFIG_DAI_NXP_SAI nxp)` you make sure that the `nxp` directory is only added to the build when the `CONFIG_DAI_NXP_SAI` symbol is selected.

When we add new NXP DAI drivers, like ESAI, we'll have:
```
add_subdirectory_ifdef(CONFIG_DAI_NXP_SAI nxp/sai)
add_subdirectory_ifdef(CONFIG_DAI_NXP_ESAI nxp/esai)
```
if you want to be consisted on how is the `dai` driver organized now - see `intel` directory.
Or, we can do:
```
zephyr_library_sources_ifdef(CONFIG_DAI_NXP_SAI sai.c)
zephyr_library_sources_ifdef(CONFIG_DAI_NXP_ESAI esai.c)
```
like other drivers directory - see for example [dma](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dma/CMakeLists.txt).

Anyway, you can look around in Zephyr, all drivers are added conditionally. :)

LaurentiuM1234 replied:
Decided to switch to Intel's approach. This way we won't have to add a `add_subdirectory_ifdef(CONFIG_DAI_NXP_SAI nxp/sai)` for every DAI driver we introduce and we maintain consistency with Intel.

LaurentiuM1234 commented on drivers/dai/nxp/sai.h:
```diff
@@ -0,0 +1,518 @@
+	 */
+	/* note: this is ignored if xCR4 doesn't have CHMOD */
+#ifdef CONFIG_SAI_HAS_XCR4_CHMOD
+	cfg->dataMode = kSAI_DataPinStateOutputZero;
```
> That's probably because only 8MP (MIMX8ML8) has CHMOD field in XCR4, right?

As well as i.MX93, yes.



> Maybe CONFIG_SAI_HAS_XCR4_CHMOD should be linked with FSL_FEATURE_SAI_HAS_CHANNEL_MODE?

This should work, yes. Will give it a try and update this PR if it works ok.

LaurentiuM1234 commented on drivers/dai/nxp/sai.h:
```diff
@@ -0,0 +1,518 @@
+		return -EPERM;
+	case DAI_STATE_READY:
+		/* TODO: for some odd reason, SOF calls
+		 * set_config() twice via the dai-legacy API.
```
Ah, this stems from the fact that when I was initially testing this driver I was just overwriting SOF's SAI driver functions with the ones from this driver, hence the usage of the legacy interface. Will re-check and see if this happens and update the comment accordingly.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+	}
+
+	bespoke = bespoke_data;
+	data = dev->data;
```
Seems like it works fine. I also do this with the new EDMA driver and no warnings either.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+		tx_config->bitClock.bclkPolarity = kSAI_PolarityActiveLow;
+		break;
+	default:
+		LOG_ERR("unsupported DAI protocol: %d",
```
I was planning on adding more protocols whenever they're required and can be tested. For now, we only support I2S and TDM as these are the only ones I worked with and actually tested (haven't actually gotten to testing TDM with the new native interface but will do so soon).

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+	tx_config->serialData.dataMaskedWord = ~bespoke->tx_slots;
+	rx_config->serialData.dataMaskedWord = ~bespoke->rx_slots;
+
+	tx_config->fifo.fifoWatermark = sai_cfg->tx_fifo_watermark - 1;
```
This is because the values from the DTS are in fact the correct (logically speaking) watermark values. You can't have a NULL watermark as that wouldn't make sense.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+#endif /* CONFIG_SAI_HAS_MCLK_CONFIG_OPTION */
+
+	/* TODO: should we leave this here? */
+	data->cfg.rate = bespoke->fsync_rate;
```
Generally speaking, `data->cfg` is used by SOF. Not sure if it does anything with the rate in particular so I'll have to double-check.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+		 * a better way of doing this? Also, is it guaranteed that
+		 * after 100US the transmitter/receiver will be disabled?
+		 */
+		k_busy_wait(100);
```
Thanks for the suggestion, will give it a shot

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+
+	LOG_DBG("pause on direction %d", dir);
+
+	ret = sai_tx_rx_disable(data, dir);
```
It seems to work for now. As far as I can tell, disabling the transmitter/receiver should also stop the DMA requests as well as the data lines from consuming/producing new data. Unless someone can make a case on why the extra operations are necessary, I want to leave `pause()` as it is for now. If there's ever any problems we'll deal with them as they appear. This way we can understand why each operation is necessary.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+	SAI_TX_RX_ENABLE_DISABLE_IRQ(dir, data->regmap,
+				     kSAI_FIFOErrorInterruptEnable, true);
+
+	/* TODO: is there a need to write some words to the FIFO to avoid starvation? */
```
This comment was left from the time I was testing the driver (using DMA domain) (although the domain we use shouldn't make a difference as the underrun would probably happen during start() if there was a problem). For now, I haven't managed to get a FIFO underrun because of this but yeah I agree with you on the testing bit.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+ * is raised.
+ *
+ * 3) Transmitter/receiver may remain enabled after sai_tx_rx_disable().
+ * Fix this.
```
Mm, what I was planning on doing with the STOP/POST_STOP operations is the following:

1. TRIG_STOP comes and sai_stop() will be called. All we need to do here is forcefully disable the transmitter and receiver.
2. TRIG_POST_STOP comes (this is processed in the next timer period so we have ~1ms) and sai_post_stop() is called. What we do here is we check if the transmitter/receiver was disabled. If not then that means we have a huge issue as this wouldn't be normal as far as I can tell. If transmitter/receiver is disabled then we proceed with the rest of the operations: line disabling and interrupt disabling.

We're probably going to end up disabling the DMA requests during STOP but we'll have to see. What I know for sure is that we're disabling the transmitter/receiver during STOP and checking if it was disabled during POST_STOP. These operations will add no delay (as we won't perform any busy waiting anywhere).

iuliana-prodan replied:
This will be solved in another PR.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+	/* check for TX FIFO error */
+	if (SAI_TX_RX_STATUS_IS_SET(DAI_DIR_TX, data->regmap, kSAI_FIFOErrorFlag)) {
+		LOG_ERR("FIFO underrun detected");
+		/* TODO: this will crash the program and should be addressed as
```
I haven't played around with the XRUN recovery so this didn't even cross my mind. I was planning on just disabling the SAI if an underrun/overrun is detected. Maybe we should consider XRUN recovery though. We'll also have to see how SOF behaves when an error occurs with the timer domain. From my experience with it, underruns/overruns don't stop the pipeline task from executing so you end up with no errors on the host side which may be a pain to deal with on the testing side of things (this is why I decided to go with `z_irq_spurious()` on underrun/overrun even though this isn't correct). We'll have to see how this plays out I guess.

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+										\
+DEVICE_DT_INST_DEFINE(inst, &sai_init, NULL,					\
+		      &sai_data_##inst, &sai_config_##inst,			\
+		      PRE_KERNEL_1, CONFIG_DAI_INIT_PRIORITY,			\
```
Is there a reason we want to change the stage at which this is initialized? Do you have a certain stage in mind?

iuliana-prodan replied:
Do we need the SAI that early, in PRE_KERNEL_1?
Also, [here](https://github.com/zephyrproject-rtos/zephyr/pull/65187/files#diff-68ee1bfebdf2de1c63b56f306a143b0ea0a3e6501d0a88967406a171dbb9c014R388) you check for pre_kernel state.
I suggest POST_KERNEL.

LaurentiuM1234 replied:
The stage check from `sai_tx_rx_disable()` will be eventually removed anyways once we split STOP into STOP and POST_STOP so we shouldn't necessarily decide based on that.

The main reason I went with `PRE_KERNEL_1` stage is because most drivers I've worked with are initialized at this stage. Also, this driver is meant to be generic so just because SOF allows a later initialization we shouldn't count on that for every use case. Despite this, I can't really give you an actual use case in which we'd prefer `PRE_KERNEL_1` instead of `POST_KERNEL` so if you still think this would be better then I wouldn't mind giving it a shot.

LaurentiuM1234 replied:
Changed to `POST_KERNEL`

LaurentiuM1234 commented on drivers/dai/nxp/sai.c:
```diff
@@ -0,0 +1,758 @@
+		/* TODO: maybe return an error code here
+		 * instead of making it seem like the
+		 * trigger operation has succeeded.
+		 */
```
This may just end up polluting the logs. For example, `dai_common_copy()` does `DAI_TRIGGER_COPY` which is not needed for us. Will have to double-check and see if this is manageable. If it's not and you think this may be helpful we could go with a `LOG_DBG` instead?

LaurentiuM1234 commented:
**V2 updates**

1. sai_trigger() now only allows `COPY` and `PRE_START` operations (apart from those which are already supported and translate to a trigger function). The rest of the trigger commands will be treated as errors. If there's ever such a trigger causing an error we're going to have to decide between adding support for it or marking it as success if the SAI doesn't need to implement it.
2. Fixed typo inside `SAI_TX_RX_SW_RESET()`.
3. `CONFIG_SAI_HAS_XCR4_CHMOD` has been entirely removed and replaced by `FSL_FEATURE_SAI_HAS_CHANNEL_MODE`.
4. Removed comment regarding READY->READY state transition (a.k.a doing a config_set after a config_set). This transition should be allowed since re-configuring the SAI immediately after a configuration isn't illogical and won't break anything. As far as I can tell, the double set_config() call from SOF still happens. Not sure why (perhaps it's done for every direction?)
5. Added explanation about why we're going to keep `data->cfg.rate = bespoke->fsync_rate;` and where it's used.
6. `sai_tx_rx_disable()` now uses `WAIT_FOR()`. The timeout value has been determined empirically. The core problem (the fac that we shouldn't be doing busy waiting) wasn't solved.

Hopefully I didn't miss anything

dbaluta commented:
Looks good to me. Now we need to merge dependencies and then we are good to go.

Thanks @LaurentiuM1234 !

mmahadevan108 commented on west.yml:
```diff
@@ -193,7 +193,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: d0c424e1c6ef0acac3099a07280a46a24951a47a
+      revision: pull/300/head
```
Please update.

LaurentiuM1234 replied:
Done

LaurentiuM1234 commented:
@kv2019i @lgirdwood could you please take a look at the changes to `dai.h`? Would like to have at least one approval from Intel's side on this. Thanks!

dbaluta commented:
All good! 

@lgirdwood @kv2019i can you please review commit modifying dai.h common part: https://github.com/zephyrproject-rtos/zephyr/pull/65187/commits/7a9401d0d045cb6d0e3b9e9e88a8572e71c25069
